File: pipeline/bin/prepare_phenotypes.R
#!/usr/bin/env Rscript
library(optparse)

option_list = list(
  make_option(c("--phenotype_file"), type="character"),
  make_option(c("--sample_file"), type="character"),
  make_option(c("--study_id"), type="character"),
  make_option(c("--group"), type="character"),
  make_option(c("--group_column"), type="character", default=""),
  make_option(c("--cases_value"), type="character", default=""),
  make_option(c("--covariate_file"), type="character"),
  make_option(c("--covariates"), type="character", default=""),
  make_option(c("--include_proteins"), type="character", default=""),
  make_option(c("--outdir"), type="character")
)


opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

dir.create(opt$outdir, showWarnings = FALSE, recursive = TRUE)
log_file <- file.path(opt$outdir, "prepare_phenotypes.log")

# Helper to read protein list
get_protein_list <- function(input) {
  if (input == "") return(NULL)
  if (file.exists(input)) {
    prots <- readLines(input, warn = FALSE)
    prots <- trimws(prots)
    prots <- prots[prots != ""]
    return(prots)
  }
  return(trimws(unlist(strsplit(input, ","))))
}

standardize_ids <- function(df) {
  cols <- colnames(df)
  target <- grep("^#IID$|^IID$|^sampleid$|^sample_id$|^SampleID$", cols, ignore.case=TRUE, value=TRUE)
  if (length(target) > 0) {
    data.table::setnames(df, target[1], "IID")
  }
  return(df)
}

# 1. Load data
pheno <- data.table::fread(opt$phenotype_file)
pheno <- standardize_ids(pheno)

if (grepl("\\.psam$", opt$sample_file)) {
    samples <- data.table::fread(opt$sample_file, skip = "#IID")
} else if (grepl("\\.fam$|\\.sam$", opt$sample_file)) {
    samples <- data.table::fread(opt$sample_file, header = FALSE)
    data.table::setnames(samples, c("FID", "IID", "PID", "MID", "SEX", "PHENO"))
} else {
    samples <- data.table::fread(opt$sample_file)
}
samples <- standardize_ids(samples)

pheno[, IID := as.character(IID)]
samples[, IID := as.character(IID)]

common_samples <- intersect(pheno$IID, samples$IID)
if (length(common_samples) == 0) stop("Zero overlap found.")

pheno <- pheno[IID %in% common_samples]

# MATCH GENETIC DATA FORMAT: Set FID to 0
pheno[, FID := "0"]
data.table::setcolorder(pheno, c("FID", "IID"))

# 2. Determine columns to keep
all_cols <- colnames(pheno)
id_cols <- c("FID", "IID")
exclude_cols <- c("STUDY", "FID", "IID", "FID_1", "IID_1") # Common metadata

# Also exclude covariates provided by user
if (opt$covariates != "") {
  user_covs <- trimws(unlist(strsplit(opt$covariates, ",")))
  exclude_cols <- unique(c(exclude_cols, user_covs))
}

req_prots <- get_protein_list(opt$include_proteins)

if (!is.null(req_prots)) {
  # If proteins are provided, keep IDs + requested proteins
  # (Still excluding anything that might be a covariate by accident)
  valid_prots <- setdiff(intersect(req_prots, all_cols), exclude_cols)
  pheno <- pheno[, c(id_cols, valid_prots), with=FALSE]
} else {
  # If NO proteins provided, keep IDs + all numeric columns (excluding metadata/covariates)
  numeric_cols <- names(which(sapply(pheno, is.numeric)))
  trait_cols <- setdiff(numeric_cols, exclude_cols)
  pheno <- pheno[, c(id_cols, trait_cols), with=FALSE]
}

# 3. Group filtering
if (!is.null(opt$group) && opt$group != "combined") {
  covs <- data.table::fread(opt$covariate_file)
  covs <- standardize_ids(covs)
  covs[, IID := as.character(IID)]

  group_df <- NULL
  if (!is.null(opt$group_column) && opt$group_column != "" && opt$group_column %in% colnames(samples)) {
    group_df <- samples[, c("IID", opt$group_column), with=FALSE]
  } else if (!is.null(opt$group_column) && opt$group_column != "" && opt$group_column %in% colnames(covs)) {
    group_df <- covs[, c("IID", opt$group_column), with=FALSE]
  } else {
    stop(paste("Group column not found in sample or covariate file:", opt$group_column))
  }

  pheno <- merge(pheno, group_df, by="IID")
  if (opt$group == "cases") pheno <- pheno[get(opt$group_column) == opt$cases_value]
  else if (opt$group == "controls") pheno <- pheno[get(opt$group_column) != opt$cases_value]
  pheno[, (opt$group_column) := NULL]
  data.table::setcolorder(pheno, c("FID", "IID"))
}

# 4. Finalise
# Using NA for quantitative traits is standard for REGENIE
# No manual replacement with -9

full_out <- file.path(opt$outdir, "full.pheno")
data.table::fwrite(pheno, full_out, sep="\t", quote=FALSE, na="NA")

keep_out <- file.path(opt$outdir, "keep_samples.txt")
data.table::fwrite(pheno[, .(FID, IID)], keep_out, sep="\t", col.names=FALSE, quote=FALSE)

# 5. Summary statistics for proteins
prot_cols <- setdiff(colnames(pheno), c("FID", "IID"))
summary_list <- list()
for (p in prot_cols) {
  vals <- pheno[[p]]
  summary_list[[p]] <- data.frame(
    protein = p,
    n_total = length(vals),
    n_missing = sum(is.na(vals)),
    mean = mean(vals, na.rm=TRUE),
    sd = sd(vals, na.rm=TRUE),
    min = min(vals, na.rm=TRUE),
    max = max(vals, na.rm=TRUE)
  )
}
summary_df <- do.call(rbind, summary_list)
data.table::fwrite(summary_df, file.path(opt$outdir, "protein_summary.tsv"), sep="\t", quote=FALSE)

File: pipeline/bin/prepare_phenotypes.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
from pathlib import Path
import re

import pandas as pd

from config import ensure_string_ids, read_path_list, read_sample_file, read_table_auto, resolve_column_name, standardize_ids


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--phenotype_file", required=True)
    parser.add_argument("--sample_file", required=True)
    parser.add_argument("--study_id", required=True)
    parser.add_argument("--group", required=True)
    parser.add_argument("--group_column", default="")
    parser.add_argument("--cases_value", default="")
    parser.add_argument("--covariate_file", required=True)
    parser.add_argument("--covariates", default="")
    parser.add_argument("--include_proteins", default="")
    parser.add_argument("--outdir", required=True)
    return parser.parse_args()


def build_feature_stem(index: int, protein_id: str) -> str:
    safe_token = re.sub(r"[^A-Za-z0-9._-]+", "_", protein_id.strip())
    safe_token = safe_token.strip("._-") or "trait"
    return f"feature_{index:05d}_{safe_token}"


def main() -> None:
    args = parse_args()
    outdir = Path(args.outdir)
    outdir.mkdir(parents=True, exist_ok=True)

    pheno = standardize_ids(read_table_auto(args.phenotype_file))
    samples = read_sample_file(args.sample_file)

    if "IID" not in pheno.columns or "IID" not in samples.columns:
        raise ValueError("Both phenotype and sample inputs must contain an IID column.")

    pheno = ensure_string_ids(pheno, ("IID",))
    samples = ensure_string_ids(samples, ("IID",))

    common_samples = pd.Index(pheno["IID"].dropna().unique()).intersection(samples["IID"].dropna().unique())
    if common_samples.empty:
        raise ValueError("Zero overlap found.")

    pheno = pheno[pheno["IID"].isin(common_samples)].copy()
    pheno.insert(0, "FID", "0")
    pheno = pheno[["FID", "IID"] + [col for col in pheno.columns if col not in {"FID", "IID"}]]

    all_columns = list(pheno.columns)
    id_columns = ["FID", "IID"]
    exclude_columns = {"STUDY", "FID", "IID", "FID_1", "IID_1"}

    if args.covariates.strip():
        exclude_columns.update(part.strip() for part in args.covariates.split(",") if part.strip())

    requested_proteins = read_path_list(args.include_proteins)

    if requested_proteins is not None:
        valid_proteins = [protein for protein in requested_proteins if protein in all_columns and protein not in exclude_columns]
        pheno = pheno[id_columns + valid_proteins].copy()
    else:
        numeric_columns = [col for col in pheno.columns if pd.api.types.is_numeric_dtype(pheno[col])]
        trait_columns = [col for col in numeric_columns if col not in exclude_columns]
        pheno = pheno[id_columns + trait_columns].copy()

    if args.group and args.group != "combined":
        covariates = standardize_ids(read_table_auto(args.covariate_file))
        if "IID" not in covariates.columns:
            raise ValueError("Covariate file must contain an IID column.")
        covariates = ensure_string_ids(covariates, ("IID",))

        sample_group_column = resolve_column_name(samples.columns, args.group_column)
        covariate_group_column = resolve_column_name(covariates.columns, args.group_column)

        if sample_group_column:
            group_frame = samples[["IID", sample_group_column]].copy()
            if sample_group_column != args.group_column:
                group_frame = group_frame.rename(columns={sample_group_column: args.group_column})
        elif covariate_group_column:
            group_frame = covariates[["IID", covariate_group_column]].copy()
            if covariate_group_column != args.group_column:
                group_frame = group_frame.rename(columns={covariate_group_column: args.group_column})
        else:
            raise ValueError(f"Group column not found in sample or covariate file: {args.group_column}")

        pheno = pheno.merge(group_frame, on="IID", how="inner", sort=False)
        group_values = pheno[args.group_column].astype("string")

        if args.group == "cases":
            pheno = pheno[group_values == str(args.cases_value)].copy()
        elif args.group == "controls":
            pheno = pheno[group_values != str(args.cases_value)].copy()

        pheno = pheno.drop(columns=[args.group_column])
        pheno = pheno[["FID", "IID"] + [col for col in pheno.columns if col not in {"FID", "IID"}]]

    full_out = outdir / "full.pheno"
    pheno.to_csv(full_out, sep="\t", index=False, na_rep="NA")

    keep_out = outdir / "keep_samples.txt"
    pheno[["FID", "IID"]].to_csv(keep_out, sep="\t", index=False, header=False)

    protein_columns = [col for col in pheno.columns if col not in {"FID", "IID"}]
    if not protein_columns:
        raise ValueError("No phenotype traits remain after filtering.")

    summary_rows = []
    manifest_rows = []

    for index, protein in enumerate(protein_columns, start=1):
        feature_out = outdir / f"{build_feature_stem(index, protein)}.pheno"
        pheno[id_columns + [protein]].to_csv(feature_out, sep="\t", index=False, na_rep="NA")
        manifest_rows.append({"protein_id": protein, "pheno_file": feature_out.name})

        values = pd.to_numeric(pheno[protein], errors="coerce")
        summary_rows.append(
            {
                "protein": protein,
                "n_total": len(values),
                "n_missing": int(values.isna().sum()),
                "mean": values.mean(),
                "sd": values.std(ddof=1),
                "min": values.min(),
                "max": values.max(),
            }
        )

    manifest_df = pd.DataFrame(manifest_rows, columns=["protein_id", "pheno_file"])
    manifest_df.to_csv(outdir / "features.manifest", sep="\t", index=False)

    summary_df = pd.DataFrame(
        summary_rows,
        columns=["protein", "n_total", "n_missing", "mean", "sd", "min", "max"],
    )
    summary_df.to_csv(outdir / "protein_summary.tsv", sep="\t", index=False)


if __name__ == "__main__":
    main()
